lib/call-cabal-project-to-nix.nix

```diff
@@ -456,6 +451,44 @@ let
     '';
   };
 
+  fakeCurl = inputMap:
+    let inputMapFile = evalPackages.writeTextFile {
+          name = "fakeCurlInputMap";
+          text = ''
+          ${builtins.concatStringsSep "\n" (pkgs.lib.mapAttrsToList f inputMap)}
+          '';
+        };
+        f = name: index: "${name}\tfile://${index}";
+    in
+      evalPackages.writeShellApplication {
+        name = "curl";
+        text = ''
+          declare -a ARGS
+          ARGS=("''${@}")
+          while read -r uri path; do
+            ARGS=("''${ARGS[@]/$uri/$path}")
+          done <${inputMapFile}
+          ${pkgs.curl}/bin/curl "''${ARGS[@]}" >/dev/null
+          echo 200
+        '';
+      };
+
+  hackageRepo = { index-state, sha256 }:
+    assert sha256 != null;
+    let at = builtins.replaceStrings [ ":" ] [ "" ] index-state;
+    in mkLocalHackageRepo {
+      name = "hackage.haskell.org";
+      index = pkgs.fetchurl {
+        name = "01-index.tar.gz-at-${at}";
+        url = "https://hackage.haskell.org/01-index.tar.gz";
+        downloadToTemp = true;
+        postFetch =
+          "${nix-tools}/bin/truncate-index -o $out -i $downloadedFile -s ${index-state}";
+        outputHashAlgo = "sha256";
+        outputHash = sha256;
+      };
+    };
```

Comment:
Here I can probably go back to use `hackageTarball`. The difference was that `hackageTarball` insists on using `name = "hackage.haskell.org-at-${at}"` while I wanted to leave `.cabal/config` untouched.

Then I had to add some fake keys in `.cabal/config` so I might as well use the different name.

Reply:
I am not so sure anymore. hackageTarball returns a attrset to describe a single thing, it looks a bit awkward to use here.